Added Python interfaces to some Ignition Gazebo methods

ahcorde opened the pull request:
Signed-off-by: ahcorde <ahcorde@gmail.com>

# 🎉 New feature

## Summary

Added Python interfaces to same Ignition Gazebo methods with pybind11

Requires https://github.com/ignitionrobotics/ign-math/pull/280

Related with this issue https://github.com/ignitionrobotics/ign-gazebo/issues/789#issuecomment-945612336

## Test it

I created an example `examples/python/helperFixture.py`.

Export your python path: 

```bash
export PYTHONPATH=$PYTHONPATH:<your workspace>/install/lib/python/
```

Run the example:

```
python3 <your workspace>/src/ign-gazebo/examples/python/testFixture.py
```

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [x] Added example and/or tutorial
- [x] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

chapulina commented on python/src/ignition/gazebo/destroyable.hh:
```diff
@@ -0,0 +1,71 @@
+{
+namespace python
+{
+/// This class blocks destruction when in use
```
Can you add some sentences explaining why this class is needed?

chapulina commented on python/src/ignition/gazebo/destroyable.hh:
```diff
@@ -0,0 +1,71 @@
+
+  /// Signal that the object should be destroyed as soon as it's not in use
+  void
+  destroy_when_not_in_use();
```
Can we use `CamelCase` for these classes, or is there a reason to use `snake_case`?

chapulina commented on python/src/ignition/gazebo/entity_component_manager.cc:
```diff
@@ -0,0 +1,57 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
```
Nit: remember to use the `/* ... */` style when cleaning up

chapulina commented on python/src/ignition/gazebo/entity_component_manager.hh:
```diff
@@ -0,0 +1,72 @@
+// limitations under the License.
+
+#ifndef IGNITION_GAZEBO_PYTHON__ENTITY_COMPONENT_MANAGER_HPP_
+#define IGNITION_GAZEBO_PYTHON__ENTITY_COMPONENT_MANAGER_HPP_
```
```suggestion
#define IGNITION_GAZEBO_PYTHON__ENTITY_COMPONENT_MANAGER_HH_
```

chapulina commented on python/src/ignition/gazebo/entity_component_manager.hh:
```diff
@@ -0,0 +1,72 @@
+
+    /// Get rcl_client_t pointer
+    ignition::gazebo::EntityComponentManager *
+    rcl_ptr() const
```
Left-overs?

ahcorde replied:
it's used in the world wrapper

chapulina commented on python/src/ignition/gazebo/event_manager.hh:
```diff
@@ -0,0 +1,39 @@
+
+#include <pybind11/pybind11.h>
+
+namespace py = pybind11;
```
I have mixed feelings about using this shorthand. We may have other python interfaces in the codebase, like SWIG, so it may be good to be explicit. Also I think `pybind11` is short enough to spell out.

chapulina commented on python/src/ignition/gazebo/exceptions.hh:
```diff
@@ -0,0 +1,35 @@
+
+namespace ignition
+{
+namespace utils
```
The `utils` namespace is reserved for `ign-utils`. Can we use `gazebo`?

chapulina commented on python/src/ignition/gazebo/helper_system.hh:
```diff
@@ -0,0 +1,159 @@
+  };
+
+/// \brief System that is inserted into the simulation loop to observe the ECM.
+class HelperSystem :
```
This looks very similar to the `HelperSystem` here:

https://github.com/ignitionrobotics/ign-gazebo/blob/96ed1efff86af0ce5d3322fdad1bf3196896e881/src/TestFixture.cc#L27

Is there a need to duplicate it? Maybe we just need to install the other one?

ahcorde replied:
it's slightly different.

I?m using the `ignition::gazebo::python::EntityComponentManager` not the `ignition::gazebo::EntityComponentManager`

chapulina replied:
Now that the `EntityComponentManager` is wrapped differently, do you think we can remove the duplication here too? Could we just use `TestFixture`?

chapulina commented on python/src/ignition/gazebo/helper_system.hh:
```diff
@@ -0,0 +1,159 @@
+{
+  class HelperFixturePrivate;
+
+  class HelperFixture : public ignition::utils::python::Destroyable, public std::enable_shared_from_this<HelperFixture>
```
Same here, this looks very similar to `TestFixture`; can we just use that class instead?

ahcorde replied:
Same comment https://github.com/ignitionrobotics/ign-gazebo/pull/1219#discussion_r756703333

chapulina commented on python/src/ignition/gazebo/server.hh:
```diff
@@ -0,0 +1,45 @@
+
+#include <memory>
+
+#include "server_config.hh"
```
I think only the `pybind11` include is needed?

chapulina commented on python/src/ignition/gazebo/world.cc:
```diff
@@ -0,0 +1,75 @@
+             std::shared_ptr<ignition::gazebo::python::World>>(module, "World")
+  .def(py::init<ignition::gazebo::Entity>())
+  .def(
+    "model_by_name", &ignition::gazebo::python::World::ModelByName,
```
It's not clear to me why we need to repeat the functions in the `python` namespace. What's preventing us from binding to `ignition::gazebo::World::ModelByName` here?

ahcorde replied:
`std::optional` can't be used as a returned type in methods which means we need to create a class to workaround this, then the rest of the methods should be wrapped too, beacuse I created a `World` instance.

chapulina replied:
Ouch, I thought pybind11 had support for C++17, including `std::optional`, and that was one of its advantages with respect to SWIG?

https://pybind11.readthedocs.io/en/stable/advanced/cast/stl.html#c-17-library-containers

_The pybind11/stl.h header also includes support for std::optional<> and std::variant<>._

ahcorde replied:
https://github.com/ignitionrobotics/ign-gazebo/pull/1219/commits/dbe9c08e435079c6d4ca47d8980a2be07f47e326

chapulina replied:
So now that we're using optional, do we still need to redefine these functions?

ahcorde replied:
every class that make use of the EntityComponentManager should be wrap

chapulina replied:
Oh that's going to be hard to scale. Is there any change we can make to the original `EntityComponentManager` class so we don't need to wrap it? Maybe we just need to make the original class copiable?

chapulina commented on examples/python/helperFixture.py:
```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/python3
```
Let's not forget to add copyright headers to all source code files

chapulina reviewed (commented):
Great work, works as advertised! At this first pass, I just have some high-level questions about the implementation.

You can ignore the style nitpicks for now and address them once we've settled on the approach.

chapulina commented on python/src/ignition/gazebo/server_config.cc:
```diff
@@ -0,0 +1,40 @@
+/*
```
Let's follow Ignition's conventions for file names, i.e.

`python/src/ignition/gazebo/ServerConfig.cc`

And so on

chapulina commented on python/src/ignition/gazebo/server_config.hh:
```diff
@@ -0,0 +1,39 @@
+ * \param[in] module a pybind11 module to add the definition to
+ */
+void
+define_server_config(pybind11::object module);
```
Can we use `camelCase` to follow our conventions?

```suggestion
void defineServerConfig(pybind11::object module);
```

chapulina reviewed (changes requested):
Just some minor stylistic comments, I think we're almost there!

It would be good to have a little test for this on CI to make sure that it's working once we release `ign-math` debs.

chapulina commented on python/src/ignition/gazebo/Exceptions.hh:
```diff
@@ -0,0 +1,38 @@
+{
+namespace python
+{
+class InvalidHandle : public std::runtime_error
```
Can't we use `std::runtime_error` directly?

chapulina replied:
Ping @ahcorde , I think we can actually get rid of this class now, it's not used anymore

chapulina commented on python/CMakeLists.txt:
```diff
@@ -0,0 +1,99 @@
+
+find_package(pybind11 2.2 QUIET)
+
+if (NOT ${pybind11_FOUND})
```
Let's only support `pybind11` from the system like we're doing on `ign-math`, otherwise it's tough to maintain multiple versions.

chapulina commented on python/CMakeLists.txt:
```diff
@@ -0,0 +1,99 @@
+
+set(PYBIND11_PYTHON_VERSION 3)
+
+find_package(pybind11 2.2 QUIET)
```
Can we move this call to the top-level `CMakeLists`? It's easier to keep track of all dependencies that way.

chapulina commented on python/CMakeLists.txt:
```diff
@@ -0,0 +1,99 @@
+    src/ignition/gazebo/EventManager.cc
+    src/ignition/gazebo/HelperSystem.cc
+    src/ignition/gazebo/Server.cc
+		src/ignition/gazebo/ServerConfig.cc
```
```suggestion
    src/ignition/gazebo/ServerConfig.cc
```

chapulina reviewed (commented):
Can we add the python dependencies to `packages.apt` so that the bindings are built on CI?

ahcorde commented:
I made some changes that might be conflictive.

 - EventManager: Changed `unique_ptr` to `shared_ptr` in a private member to be able to copy the class
 - TextureFixture: Changed some reference to pointers
    - some integration tests have some changes: Spherical_coordinates and thruster
 
I can open a different PRs, these PRs might require to `tick-tock` some of the methods

chapulina reviewed (commented):
> I made some changes that might be conflictive.
>
>    EventManager: Changed unique_ptr to shared_ptr in a private member to be able to copy the class
>    TextureFixture: Changed some reference to pointers
>        some integration tests have some changes: Spherical_coordinates and thruster
>
> I can open a different PRs, these PRs might require to tick-tock some of the methods

Ah yeah those would need to be tick-tocked. The `EventManager` one is less concerning, but I'd like to avoid the `TestFixture` one because that breaks the API contract that we have of making the ECM immutable in some callbacks, and dis-encouraging callbacks from storing that pointer.

Is there something about pybind11 that prevents passing a mutable reference as an argument? I know that on `ign-math` we're [returning tuples](https://github.com/ignitionrobotics/ign-math/blob/ign-math6/src/python_pybind11/src/Quaternion.hh#L127-L133) in some cases to work around that, but reading about pybind11's [limitations involving reference arguments](https://pybind11.readthedocs.io/en/stable/faq.html#limitations-involving-reference-arguments), it looks like that may only be necessary for POD?

ahcorde commented on python/CMakeLists.txt:
```diff
@@ -0,0 +1,83 @@
+)
+
+configure_build_install_location(gazebo)
+configure_build_install_location(common)
```
it's installed in the lines above. the `gazebo` and `common` module are installed under the `ignition` folder to be able to import them with `import ignition.gazebo` or `import ignition.common`, but `sdformat` I decided to just imported like: `import sdformat`. 

What do you think?

chapulina replied:
Ah nevermind, I missed that

j-rivero replied:
I think that using that approach is just fine but we are losing the logic of https://github.com/ignitionrobotics/ign-gazebo/pull/1219/files#diff-005a07ca82fb46b215f58a3a8f4d6d6cc6ca088ade44e8f15cc2e6a51d1da6f4R9-R29 when computing the install path for the sdformat and go with a simple `DESTINATION "${IGN_LIB_INSTALL_DIR}/python/"` which is not working for bindings packaging. We need to preserve the logic and add the `ignition` extra dir when needed.

j-rivero replied:
Something like ...
```diff
diff --git a/python/CMakeLists.txt b/python/CMakeLists.txt
index bc98f79c..7dedff0a 100644
--- a/python/CMakeLists.txt
+++ b/python/CMakeLists.txt
@@ -28,13 +28,11 @@ else()
   set(IGN_PYTHON_INSTALL_PATH ${IGN_LIB_INSTALL_DIR}/python)
 endif()
 
-set(IGN_PYTHON_INSTALL_PATH "${IGN_PYTHON_INSTALL_PATH}/ignition")
-
 # Set the build location and install location for a CPython extension
 function(configure_build_install_location _library_name)
   # Install library for actual use
   install(TARGETS ${_library_name}
-    DESTINATION "${IGN_PYTHON_INSTALL_PATH}/"
+    DESTINATION "${IGN_PYTHON_INSTALL_PATH}/ignition"
   )
 endfunction()
 
@@ -77,7 +75,7 @@ target_link_libraries(sdformat PRIVATE
 )
 
 install(TARGETS sdformat
-       DESTINATION "${IGN_LIB_INSTALL_DIR}/python/"
+  DESTINATION "${IGN_PYTHON_INSTALL_PATH}"
 )
 
 configure_build_install_location(gazebo)
```

chapulina commented on examples/python/helperFixture.py:
```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/python3
```
Let's rename this file to test fixture

chapulina commented on python/src/ignition/gazebo/World.cc:
```diff
@@ -0,0 +1,45 @@
+    " this world.")
+  .def(
+    "gravity", &ignition::gazebo::World::Gravity,
+    "Get the gravity in m/s^2.");
```
Ping @ahcorde , this is still happening to me, so I can't fully test the PR. Doesn't it happen to you?

chapulina replied:
While debugging the `unique_ptr` usage, I noticed that changing back to `unique_ptr` and commenting out `helper.finalize()` at least allows me to run the example without runtime errors.

chapulina replied:
I can't reproduce it anymore! :tada:

chapulina commented on include/ignition/gazebo/EntityComponentManager.hh:
```diff
@@ -768,8 +768,10 @@ namespace ignition
           const std::unordered_set<ComponentTypeId> &_types = {}) const;
 
+      /// This was originally a unique_ptr, but pybind11 was failing because it
+      /// was not able to copy the class
```
I think this is a symptom of a bigger issue. Nowhere on the example we should be trying to copy the ECM. I think we need to understand where this is happening.

I tried changing it back to `unique_ptr` and confirmed that the example fails with:

```
Traceback (most recent call last):                          
  File "src/ign-gazebo/examples/python/helperFixture.py", line 64, in <module>
    helper.finalize()
RuntimeError: return_value_policy = copy, but the object is non-copyable!
```

I traced it back to this line:

https://github.com/ignitionrobotics/ign-gazebo/blob/ae36d833f01e7078ff8bb397bee5981a0e93ed13/src/TestFixture.cc#L165

If I comment it out the example runs (but it's incorrect as expected). I'm still trying to make sense of it.

chapulina replied:
I see you reverted the pointer change, great! Let's also remove the comment.

chapulina commented on examples/python/testFixture.py:
```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/python3
```
Let's add a short README do this directory explaining how to run the example. Just for people who land on the directory directly and don't know about the tutorial.

chapulina commented on include/ignition/gazebo/EventManager.hh:
```diff
@@ -140,9 +140,11 @@ namespace ignition
       /// \brief Container of used signals.
       private: std::unordered_map<TypeInfoRef,
-                                  std::unique_ptr<ignition::common::Event>,
+                                  std::shared_ptr<ignition::common::Event>,
```
Can we revert this change like you did for the ECM? Can the same logic be applied here? Otherwise we'll need to retarget the PR to Garden because this breaks ABI.

chapulina commented on tutorials/python_interfaces.md:
```diff
@@ -0,0 +1,79 @@
+
+Ignition Gazebo provides a Python API to interact with world.
+
+For now, we provide a `HelperFixture` class that allows to load a world file,
```
```suggestion
For now, we provide a `TestFixture` class that allows to load a world file,
```

More updates are needed below

chapulina reviewed (changes requested):
This is looking great! It's amazing how you find a solution for all the problems jejeje

I think the only outstanding big issue is the ABI break on `EventManager`. We also need an `ign-math` release to make sure the test passes.

chapulina reviewed (changes requested):
It's looking great! I just have final suggestions in https://github.com/ignitionrobotics/ign-gazebo/pull/1308 and after that I'm ready to approve! :tada:

chapulina reviewed (approved):
Yay! This is great! Now we just need an `ign-math6` stable release for the tests to pass.

chapulina commented:
I tested this branch against `ign-math6.9.3~pre2` debs, and it works well! :tada:

chapulina commented:
A couple of problems in Jenkins CI.

1. The test fails (but it passes on GitHub actions)

```
213: Test command: /usr/bin/python3 "/home/jenkins/workspace/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/ign-gazebo/python/test/testFixture_TEST.py"
213: Environment variables: 
213:  PYTHONPATH=/usr/lib/python/
213:  LD_LIBRARY_PATH=/usr/lib:/usr/local/lib/x86_64-linux-gnu
213: Test timeout computed to be: 1500
213: Traceback (most recent call last):
213:   File "/home/jenkins/workspace/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/ign-gazebo/python/test/testFixture_TEST.py", line 19, in <module>
213:     from ignition.common import set_verbosity
213: ModuleNotFoundError: No module named 'ignition.common'
213/213 Test #213: testFixture_TEST.py ....................................***Failed    0.23 sec
```

2. Jenkins doesn't report the failure (CC @j-rivero )

chapulina commented:
For packaging, we'll need to do the same as https://github.com/ignition-release/ign-math6-release/pull/6 for `ign-gazebo6-release`. @j-rivero said he can help in the coming days. The plan is to install the `ign-common` and `sdformat` bindings as if they came from those packages, and if we ever move the bindings there, we should be able to flip the switch without affecting users.

j-rivero reviewed (approved):
Patch seems to work as expected, [tested](https://build.osrfoundation.org/job/ign-gazebo6-debbuilder/549/consoleFull#console-section-7) via patch in release repo here:
```
-- Installing: /home/jenkins/workspace/ign-gazebo6-debbuilder/build/ignition-gazebo-6.6.0~pre1/debian/tmp/usr/lib/python3/dist-packages/sdformat.cpython-38-x86_64-linux-gnu.so
-- Installing: /home/jenkins/workspace/ign-gazebo6-debbuilder/build/ignition-gazebo-6.6.0~pre1/debian/tmp/usr/lib/python3/dist-packages/ignition/gazebo.cpython-38-x86_64-linux-gnu.so
-- Set runtime path of "/home/jenkins/workspace/ign-gazebo6-debbuilder/build/ignition-gazebo-6.6.0~pre1/debian/tmp/usr/lib/python3/dist-packages/ignition/gazebo.cpython-38-x86_64-linux-gnu.so" to ""
```

osrf-triage commented:
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-ignition-releases-2022-03-25-fortress-edifice-citadel/1343/1
